### PR TITLE
L2 norm regularization term to the loss to prevent gradient exp…

### PR DESCRIPTION
**Pull request by r-zemblys**
Reopened this instead of [#75](https://github.com/ibab/tensorflow-wavenet/pull/75).
FIX: return total loss. Thanks @jyegerlehner 
FIX: default C is now 0. Thanks @jyegerlehner 
TODO: @jyegerlehner : C looks like something we would want to configure from command line or wavenet_params.json.

### CODE REVIEW

**Comment by lemonzi** [comment]
The biases shouldn't be regularized: http://stats.stackexchange.com/questions/86991/reason-for-not-shrinking-the-bias-intercept-term-in-regression.

File: wavenet.py [hunk 1]
```diff
@@ -210,7 +210,7 @@ def predict_proba(self, waveform, name='wavenet'):
                             [1, self.quantization_channels])
             return tf.reshape(last, [-1])
 
-    def loss(self, input_batch, name='wavenet'):
+    def loss(self, input_batch, C=0., name='wavenet'):
```
Comment (ibab) [inline wavenet.py]: The `C` parameter could use a more descriptive name, like `regularization_strength`, `regularization_coefficient`, or just `regularization`.
We should also set it to `None` by default, and not create any of the regularization ops below if it's `None`.
That way we the user doesn't anything to the TensorFlow graph when they don't want to use regularization.

File: train.py [hunk 2]
```diff
@@ -27,7 +27,7 @@
 WAVENET_PARAMS = './wavenet_params.json'
 STARTED_DATESTRING = "{0:%Y-%m-%dT%H-%M-%S}".format(datetime.now())
 SAMPLE_SIZE = 100000
-
+L2_REGULARIZATION_STRENGTH=0.
```
Comment (ibab) [inline train.py]: It would be nice to put spaces around the `=` and to set the parameter equal to `None` by default.

File: wavenet.py [hunk 3]
```diff
@@ -234,7 +234,16 @@ def loss(self, input_batch, name='wavenet'):
                     prediction,
                     tf.reshape(shifted, [-1, self.quantization_channels]))
                 reduced_loss = tf.reduce_mean(loss)
-
+
```
Comment (ibab) [inline wavenet.py]: There are some extra spaces in this line that can be removed.

File: wavenet.py [hunk 4]
```diff
@@ -234,7 +234,16 @@ def loss(self, input_batch, name='wavenet'):
                     prediction,
                     tf.reshape(shifted, [-1, self.quantization_channels]))
                 reduced_loss = tf.reduce_mean(loss)
-
+                
+                # L2 regularization for all trainable parameters
+                l2_loss = tf.add_n([tf.nn.l2_loss(v) 
+                                    for v in tf.trainable_variables()])
+
```
Comment (ibab) [inline wavenet.py]: extra spaces in this line.

File: wavenet.py [hunk 5]
```diff
@@ -234,7 +234,16 @@ def loss(self, input_batch, name='wavenet'):
...
                 reduced_loss = tf.reduce_mean(loss)
-
+                
+                # L2 regularization for all trainable parameters
+                l2_loss = tf.add_n([tf.nn.l2_loss(v) 
+                                    for v in tf.trainable_variables()])
+                                        
+                # Add the regularization term to the loss
+                total_loss = reduced_loss + C * l2_loss
+
```
Comment (ibab) [inline wavenet.py]: extra spaces in this line.

File: wavenet.py [hunk 6]
```diff
@@ -234,7 +234,17 @@ def loss(self, input_batch, name='wavenet'):
                     prediction,
                     tf.reshape(shifted, [-1, self.quantization_channels]))
                 reduced_loss = tf.reduce_mean(loss)
-
+                
+                # L2 regularization for all trainable parameters
+                l2_loss = tf.add_n([tf.nn.l2_loss(v) 
+                                    for v in tf.trainable_variables() 
+                                    if not('bias' in v.name)])
+
```
Comment (ibab) [inline wavenet.py]: extra spaces here.

File: wavenet.py [hunk 7]
```diff
@@ -234,7 +234,17 @@ def loss(self, input_batch, name='wavenet'):
...
-
+                
+                # L2 regularization for all trainable parameters
+                l2_loss = tf.add_n([tf.nn.l2_loss(v) 
+                                    for v in tf.trainable_variables() 
+                                    if not('bias' in v.name)])
+                
+                # Add the regularization term to the loss
+                total_loss = reduced_loss + C * l2_loss
+
```
Comment (ibab) [inline wavenet.py]: extra spaces here.

File: train.py [hunk 8]
```diff
@@ -203,7 +206,7 @@ def main():
     sess = tf.Session(config=tf.ConfigProto(log_device_placement=False))
     init = tf.initialize_all_variables()
     sess.run(init)
-
+
```
Comment (lemonzi) [inline train.py]: Trailing space.

File: train.py (hunk 2 above)
Comment (lemonzi) [inline train.py]: Should this be part of `wavenet_params.json`? It doesn't change how generation works, but it will result in very different models.

Comment (ibab) [inline train.py]: I can imagine that users might want to stop the training and restart it with a different regularization parameter. Although this would also work by editing `wavenet_params.json`, I think that having it as part of the training script makes things easier.
It also seems natural to have it next to the learning rate.

Comment (lemonzi) [inline train.py]: Then, it might make sense to have a second JSON file with training parameters: L2 regularization, learning rate, learning rate decay, dropout rate, ...

Comment (lemonzi) [inline train.py]: Oh, and it should also be easy to automatically add a flag for each JSON field allowing manual overrides.

Comment (ibab) [inline train.py]: I like the idea of a second JSON file.
Maybe there are already some Python libraries out there that handle this for us?

Comment (lemonzi) [inline train.py]: The easy way is to define everything using flags and argparse, and then supply besides the flags a list of JSON files. The JSON files will be sequentially read and merged with the current set of options, and finally the dictionary from argparse will be merged with the current set, overriding everything. I can try and make a PR with this. I don't know of any libraries that may make this easier. The question then is whether the flags are global or we allow nesting via a magic nesting character in the flag name like `/` or `.`.

We could also define flags at different places (each file defines the global flag it needs), which works well with large codebases, but maybe the repo is not large enough for this.

Comment (lemonzi) [inline train.py]: I'll open an issue for this, for the time being the regularization can stay as a separate flag.

Comment (jyegerlehner) [inline train.py]: > I like the idea of a second JSON file.

I dislike the idea of a second JSON file.

> a second JSON file with training parameters

Weren't you the one complaining about too many hyperparameters? Well I'm going to complain about too many config files. How about just putting two objects within a single file, one for model parameters and a second for training params.

Comment (ibab) [inline train.py]: The reason I was thinking of splitting them up is that we're currently using the `wavenet_params.json` file to encode the dimensions of the network. The file needs to stay the same between training and generation.
That's why I'm a bit reluctant to put training parameters in there as well, but I agree that we need a place to put the parameters.
Rest assured that we'll try to find a sane solution, which may also just involve putting everything into a single file. The discussion on this is in #79 and #80.

**Comment by ibab** [comment]
You've changed the `None` default back to `0`.
Was there a problem with using `None`?

Also, there's a conflict with `train.py`.
If you want, I can take your changes and apply them to the project (your name will appear on the commit).

**Comment by r-zemblys** [comment]
There was a problem setting `None` as default in `argparse`, so I just came back to 0. 
Yes please, resolve that conflict.

**Comment by ibab** [comment]
I've added your changes to #88, so they will end up on `master` once it's merged. Thank you!
